### ISSUE TITLE
How train with ArcFaceLoss ?

### ISSUE THREAD

**Issue by bouachalazhar** [KevinMusgrave/pytorch-metric-learning#534]
I would like to train a model but I have a problem when I create a training where add this code :
```
loss_func = losses.ArcFaceLoss(...).to(torch.device('cuda'))
loss_optimizer = torch.optim.SGD(loss_func.parameters(), lr=0.01)
# then during training:
loss_optimizer.step()
```
In this loop :
```
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
loss_func = losses.ArcFaceLoss(...).to(torch.device('cuda'))
loss_optimizer = torch.optim.SGD(loss_func.parameters(), lr=0.01)
for epoch_id in range(epochs):
    for iter_id, batch in tqdm(enumerate(dataloaders.get('train'))):
        images = batch[0].to(device)
        labels = batch[1].to(device)
        embeddings = model(images)
        loss = loss_func(embeddings, labels)
        optimizer.zero_grad()
        loss_optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        loss_optimizer.step()
```

And I would like to add scheduler with Arcfaceloss but I don't know where add this : 
```
scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
iters = len(dataloader)
for epoch in range(20):
    for i, sample in enumerate(dataloader):
        inputs, labels = sample['inputs'], sample['labels']
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = loss_func(outputs, labels)
        loss.backward()
        optimizer.step()
        scheduler.step(epoch + i / iters)
```

**Comment by KevinMusgrave** [comment]
Could you be more specific? Are you getting an error? If so please paste in the error.

**Comment by bouachalazhar** [comment]
I think Arcfaceloss doesn't work you say add this : 
`loss_optimizer.step()
`
But I don't know where

**Comment by KevinMusgrave** [comment]
Put that in the for loop, along with the other `optimizer.step()`. I think you've already done that. Why don't you think it works?

**Comment by bouachalazhar** [comment]
I'm not sure this code is correct : 
```
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
loss_func = losses.ArcFaceLoss(...).to(torch.device('cuda'))
loss_optimizer = torch.optim.SGD(loss_func.parameters(), lr=0.01)
for epoch_id in range(epochs):
    for iter_id, batch in tqdm(enumerate(dataloaders.get('train'))):
        images = batch[0].to(device)
        labels = batch[1].to(device)
        embeddings = model(images)
        loss = loss_func(embeddings, labels)
        optimizer.zero_grad()
        loss_optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        loss_optimizer.step()
```
And I would like to add a `torch.optim.lr_scheduler.CosineAnnealingWarmRestarts` but I have two scheduler I think it isn't ok : 
```
scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
scheduler2 = CosineAnnealingWarmRestarts(loss_optimizer, T_0, T_mult)
iters = len(dataloader)
for epoch in range(20):
    for i, sample in enumerate(dataloader):
        inputs, labels = sample['inputs'], sample['labels']
        optimizer.zero_grad()
        loss_optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        loss_optimizer.step()
        scheduler.step(epoch + i / iters)
        scheduler2.step(epoch + i / iters)
```

**Comment by KevinMusgrave** [comment]
It looks fine to me. If you want to simplify the code, you can use the same optimizer for the model + loss function:

```python
optimizer = torch.optim.SGD(model.parameters() + loss_func.parameters(), lr=0.01)
```

**Comment by bouachalazhar** [comment]
Like this now I have one `scheduler` : 
```
optimizer = torch.optim.SGD(model.parameters() + loss_func.parameters(), lr=0.01)
scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
iters = len(dataloader)
for epoch in range(20):
    for i, sample in enumerate(dataloader):
        inputs, labels = sample['inputs'], sample['labels']
        optimizer.zero_grad()
        loss_optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        loss_optimizer.step()
        scheduler.step(epoch + i / iters)
```

**Comment by KevinMusgrave** [comment]
You can get rid of `loss_optimizer.step()` and `loss_optimizer.zero_grad()`. Otherwise it looks fine.

**Comment by bouachalazhar** [comment]
Yes, I forgot : 
```
optimizer = torch.optim.SGD(model.parameters() + loss_func.parameters(), lr=0.01)
scheduler = CosineAnnealingWarmRestarts(optimizer, T_0, T_mult)
iters = len(dataloader)
for epoch in range(20):
    for i, sample in enumerate(dataloader):
        inputs, labels = sample['inputs'], sample['labels']
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        scheduler.step(epoch + i / iters)
```
Thanks for your help, I have a question for train with ArcFaceLoss can I use `torch.optim.Adam` ?

**Comment by KevinMusgrave** [comment]
Yes you can use Adam with ArcFace

**Comment by bouachalazhar** [comment]
My last question is it a good practice to choose the same learning_rate for model.parameters() and loss_func.parameters() ?

**Comment by KevinMusgrave** [comment]
Yeah that's fine. If you're using a pretrained model, you could set the `loss_func` learning rate higher. For example you could set it to be 10x the lr of the model's lr.

**Comment by bouachalazhar** [comment]
Ok but if I have this `optimizer = torch.optim.SGD(model.parameters() + loss_func.parameters(), lr=0.01)` how define 10xlearning_rate_model, when learning_rate_model : model's learning_rate

**Comment by KevinMusgrave** [comment]
You can pass in a list of dictionaries to the optimizer. See https://pytorch.org/docs/stable/optim.html#per-parameter-options

**Comment by bouachalazhar** [comment]
Amazing ! For me if I take 10xlearning_rate_model : 
```
optimizer = torch.optim.SGD([ {'params': model.parameters()},
                              {'params': loss_func.parameters(), 'lr': 10*1e-3} ], 
                              lr=1e-3, momentum=0.9)
```
Thanks for all !
